### PR TITLE
Inline execution event data can be offloaded to cloud blobstore 

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Inline event data can now configured to be offloaded rather than saved in the flyteadmin db.
In the future we can explore an async write approach for this code path.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/1789

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/runtime/interfaces/application_configuration.go [hunk 1]
```diff
@@ -301,7 +301,18 @@ type SignedURL struct {
 	SigningPrincipal string `json:"signingPrincipal"`
 }
 
-// This configuration handles all requests to get remote data such as execution inputs & outputs.
+type InlineEventDataPolicy = string
+
+const (
+	// InlineEventDataPolicyOffload specifies that inline execution event data (e.g. outputs) should be offloaded to the
+	// configured cloud blob store.
+	InlineEventDataPolicyOffload = "offload"
```
Comment (EngHabu) [inline pkg/runtime/interfaces/application_configuration.go:309]:
```suggestion
	InlineEventDataPolicyOffload InlineEventDataPolicy = iota
```

File: pkg/runtime/interfaces/application_configuration.go [hunk 2]
```diff
@@ -301,7 +301,18 @@ type SignedURL struct {
...
+	// InlineEventDataPolicyStoreInline specifies that inline execution event data should be saved inline with execution
+	// database entries.
+	InlineEventDataPolicyStoreInline = "inline"
```
Comment (EngHabu) [inline pkg/runtime/interfaces/application_configuration.go:312]:
```suggestion
	InlineEventDataPolicyStoreInline
```

File: pkg/runtime/interfaces/application_configuration.go [hunk 3]
```diff
@@ -301,7 +301,18 @@ type SignedURL struct {
 	SigningPrincipal string `json:"signingPrincipal"`
 }
 
-// This configuration handles all requests to get remote data such as execution inputs & outputs.
+type InlineEventDataPolicy = string
```
Comment (EngHabu) [inline pkg/runtime/interfaces/application_configuration.go:304]: //go:generate enumer -type=InlineEventDataPolicy -trimprefix=InlineEventDataPolicy

type InlineEventDataPolicy int

File: pkg/runtime/interfaces/application_configuration.go [hunk 4]
```diff
@@ -311,6 +322,8 @@ type RemoteDataConfig struct {
 	SignedURL SignedURL `json:"signedUrls"`
 	// Specifies the max size in bytes for which execution data such as inputs and outputs will be populated in line.
 	MaxSizeInBytes int64 `json:"maxSizeInBytes"`
+	// Specifies how inline execution event data should be saved in the backend
+	InlineEventDataPolicy InlineEventDataPolicy `json:"inlineEventDataPolicy"`
```
Comment (EngHabu) [inline pkg/runtime/interfaces/application_configuration.go:326]:
```suggestion
	InlineEventDataPolicy InlineEventDataPolicy `json:"inlineEventDataPolicy" pflag:",Specifies how inline execution event data should be saved in the backend"`
```

File: pkg/repositories/transformers/constants.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package transformers
+
+const Outputs = "offloaded_outputs"
```
Comment (EngHabu) [inline pkg/repositories/transformers/constants.go:3]: What does this `Outputs` const represent? maybe a slightly more descriptive name helps? or a comment line?
something like this?
```suggestion
const OutputsObjectSuffix = "offloaded_outputs"
```

File: pkg/runtime/interfaces/application_configuration.go (hunk 1 above)
Comment (katrogan) [reply pkg/runtime/interfaces/application_configuration.go:309]: how do you users specify this in the config if we're using iota?

File: pkg/repositories/transformers/constants.go (hunk 5 above)
Comment (katrogan) [reply pkg/repositories/transformers/constants.go]: done

File: pkg/runtime/interfaces/application_configuration.go (hunk 4 above)
Comment (katrogan) [reply pkg/runtime/interfaces/application_configuration.go]: done

**Comment by katrogan** [comment]
PTAL @EngHabu

File: pkg/runtime/interfaces/application_configuration.go (hunk 1 above)
Comment (EngHabu) [reply pkg/runtime/interfaces/application_configuration.go:309]: they can still say `offload` and `inline`... if you generate enumer that generates json marshal and unmarshal and stdlib/config find that automatically and use it....

**Comment by kumare3** [comment]
>> In the future we can explore an async write approach for this code path

I do not think this is possible with the durability guarantees. We may use potentially a faster more expensive store

File: pkg/runtime/interfaces/application_configuration.go (hunk 1 above)
Comment (katrogan) [reply pkg/runtime/interfaces/application_configuration.go]: ohhh I see, thanks. done!

**Comment by katrogan** [comment]
PTAL @EngHabu

**Review by EngHabu — DISMISSED** [review]

**Comment by katrogan** [comment]
@EngHabu sorry, one more time? (had merge conflicts)

**Comment by katrogan** [comment]
PTAL @EngHabu

**Review by EngHabu — APPROVED** [review]
